cli/src/cli_util.rs

```diff
@@ -1362,6 +1375,96 @@ Then run `jj squash` to move the resolution into the conflicted commit."#,
 
         Ok(())
     }
+
+    // Retrieves pattern literals from the user's advance-branches config and parses
+    // them.
+    fn advance_branches_patterns(
+        &self,
+        setting_key: &str,
+    ) -> Result<Vec<StringPattern>, CommandError> {
+        let setting = format!("experimental-advance-branches.{setting_key}");
+        let err_msg = format!("Error parsing pattern for {setting}");
+        match self.settings.config().get_array(&setting) {
+            Ok(patterns) => patterns
+                .into_iter()
+                .map(|p| {
+                    p.into_string()
+                        .map_err(|e| config_error_with_message(&err_msg, e))
+                        .and_then(|s| {
+                            StringPattern::parse(&s)
+                                .map_err(|e| config_error_with_message(&err_msg, e))
+                        })
+                })
+                .collect(),
+            Err(_) => Ok(Vec::new()),
+        }
+    }
+
+    // Returns a set of patterns that can be used to identify local branches
+    // which have the advance-branches setting enabled.
+    fn advance_branches_enabled_branches(&self) -> Result<Vec<StringPattern>, CommandError> {
+        self.advance_branches_patterns("enabled-branches")
+    }
+
+    // Returns a set of patterns that can be used to identify local branches
+    // which have the advance-branches setting disabled.
+    fn advance_branches_disabled_branches(&self) -> Result<Vec<StringPattern>, CommandError> {
+        self.advance_branches_patterns("disabled-branches")
+    }
+
+    /// Identifies branches which are eligible to be moved automatically during
+    /// `jj commit` and `jj new`. Whether a branch is eligible is determined by
+    /// its target and the user and repo config for "advance-branches".
+    ///
+    /// Returns a Vec of branches in `repo` that point to any of the `from`
+    /// commits and that are eligible to advance. The `from` commits are
+    /// typically the parents of the target commit of `jj commit` or `jj new`.
+    ///
+    /// Branches are not moved until
+    /// `WorkspaceCommandTransaction::advance_branches()` is called with the
+    /// `AdvanceableBranch`s returned by this function.
+    ///
+    /// Returns an empty `std::Vec` if no branches are eligible to advance.
+    pub fn get_advanceable_branches<'a>(
+        &self,
+        from: impl IntoIterator<Item = &'a CommitId>,
+    ) -> Result<Vec<AdvanceableBranch>, CommandError> {
+        let enabled_branches = self.advance_branches_enabled_branches()?;
+        let disabled_branches = self.advance_branches_disabled_branches()?;
```

Comment:
nit: might be better to add a separate `AdvanceBranchesSettings`/`Patterns` type and move parsing functions there. `WorkspaceCommandHelper` tends to get messier.

```rust
struct AdvanceBranchesSettings {
    enabled_branches,
    disabled_branches,
}

impl AdvancedBranchesSettings {
    fn from_config(config) -> Result<Self, CommandError>
}
```

Reply:
Yeah, good idea. I didn't really like how this cluttered up the workspace helper either.